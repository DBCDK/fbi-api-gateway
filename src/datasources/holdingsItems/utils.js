// Helper function to construct the URL dynamically
export function buildPath(
  url,
  { agencyId, bibliographicRecordId, trackingId, itemId }
) {
  const basePath = [url, agencyId, bibliographicRecordId, itemId]
    .filter(Boolean)
    .join("/");
  const query = trackingId
    ? `?trackingId=${encodeURIComponent(trackingId)}`
    : "";
  return `${basePath}${query}`;
}
