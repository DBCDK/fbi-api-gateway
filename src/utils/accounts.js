/**
 * @file This file handles util function to select single or multiple CULR accounts
 *
 */

import { isValidCpr } from "./cpr";

/**
 *
 * @param {string} accessToken
 * @param {object} context
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Promise} Object or null
 */
export async function getAccount(accessToken, context, props) {
  return (await getAccounts(accessToken, context, props))?.[0] || null;
}

/**
 *
 * @param {string} accessToken
 * @param {object} context
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Promise} (Array)
 */
export async function getAccounts(accessToken, context, props) {
  // userInfo
  const userinfo = await context.datasources.getLoader("userinfo").load({
    accessToken,
  });

  const user = userinfo?.attributes;

  if (!user?.userId) {
    return [];
  }

  // select dataloader
  let dataloader = isValidCpr(user.userId)
    ? "culrGetAccountsByGlobalId"
    : "culrGetAccountsByLocalId";

  // Retrieve user culr account
  const response = await context.datasources.getLoader(dataloader).load({
    userId: user.userId,
    agencyId: user.loggedInAgencyId,
  });

  return filterAccountsByProps(response.accounts, props);
}

/**
 *
 * @param {Array} accounts
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Array}
 */
export function filterAccountsByProps(accounts = [], props = {}) {
  if (props.agency) {
    accounts = accounts?.filter(({ agencyId }) => agencyId === props.agency);
  }
  if (props.id) {
    accounts = accounts?.filter(({ userIdValue }) => userIdValue === props.id);
  }
  if (props.type) {
    accounts = accounts?.filter(({ userIdType }) => userIdType === props.type);
  }
  return accounts;
}

/**
 *
 * Function to select a specific agency from a userinfo agencies list
 *
 * @param {Array} agencies
 * @param {object} props - Filter agencies by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Array}
 */
export function filterAgenciesByProps(agencies = [], props = {}) {
  if (props.agency) {
    agencies = agencies?.filter(({ agencyId }) => agencyId === props.agency);
  }
  if (props.id) {
    agencies = agencies?.filter(({ userId }) => userId === props.id);
  }
  if (props.type) {
    agencies = agencies?.filter(({ userIdType }) => userIdType === props.type);
  }
  return agencies;
}
