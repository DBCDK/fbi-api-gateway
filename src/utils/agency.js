/**
 * Function to check if an agencyId is a FFU library
 *
 * @param {string} branchId
 *
 * @returns {boolean}
 */

export function _isFFUAgency(branchId) {
  const LENGTH = 6;
  const list = ["4", "6", "8", "9"];
  return branchId.length === LENGTH && list.includes(branchId.charAt(0));
}

/**
 * Function to check if an agencyId is a FFU library
 *
 * @param {string} branchId
 * @param {object} context
 *
 * @returns {boolean}
 */

export async function isFFUAgency(branchId, context) {
  if (!context) {
    // If no context given/available, fallback to old/static check.
    return _isFFUAgency(branchId);
  }

  const loader = context?.getLoader || context?.datasources?.getLoader;

  const result = (await loader("library").load({ branchId })).result?.[0];

  // toUpperCase needed for mocked agencies (jest testing)
  return !!(result?.agencyType?.toUpperCase() === "FORSKNINGSBIBLIOTEK");
}

/**
 *
 * @param {string} branchId
 * @param {object} context
 *
 * @returns {boolean}
 */

export async function isFolkAgency(branchId, context) {
  if (!context) {
    // If no context given/available, fallback to old/static check.
    return branchId.startsWith("7");
  }

  const loader = context?.getLoader || context?.datasources?.getLoader;

  const result = (await loader("library").load({ branchId })).result?.[0];

  // toUpperCase needed for mocked agencies (jest testing)
  return !!(result?.agencyType?.toUpperCase() === "FOLKEBIBLIOTEK");
}

/**
 * Function to check if an agency sync data with culr
 *
 * @param {string} branchId
 * @returns {boolean}
 */
export async function hasCulrDataSync(branchId, context) {
  /**
   * Odense Katedralskole, 872960,
   * Roskilde Gymnasium, 872600
   * Sorø Akademis Skole, 861640
   * Slagelse Gymnasium, biblioteket 872320
   * Greve Gymnasium, Biblioteket 874260
   * Stenhus Gymnasium, 875140
   * Sct. Knuds Gymnasium, 871890
   */

  const whitelist = [
    // Gymnasier
    "872960",
    "872600",
    "861640",
    "872320",
    "874260",
    "875140",
    "871890",

    // Login with mitId, when no library accounts
    "190101",
  ];

  if (whitelist.includes(branchId)) {
    return true;
  }

  return !!(await isFolkAgency(branchId, context));
}

/**
 * Function to check if an branch acts as independent agency
 *
 * @param {string} branchId // agencyId or branchId
 * @returns {boolean}
 */
export function branchIsIndependent(branchId) {
  const whitelist = [
    // Agencies
    "876040", // Nordjyske Gymnasiebiblioteker agency

    // Branches
    "872050", // Støvring Gymnasium, Biblioteket
    "872060", // Hjørring Gymnasium og HF-kursus, Biblioteket
    "872080", // Hasseris Gymnasium, Biblioteket
    "872090", // Brønderslev Gymnasium og HF-Kursus, Biblioteket
    "872100", // Vesthimmerlands Gymnasium & HF, Biblioteket
    "872140", // Aalborghus Gymnasium, Biblioteket
    "872340", // Frederikshavn Gymnasium, Gymnasiebiblioteket
    "872520", // Aalborg Katedralskole, Biblioteket
    "873310", // Nørresundby Gymnasium og HF, Biblioteket
    "874100", // Dronninglund Gymnasium, Biblioteket
  ];

  return whitelist.includes(branchId);
}

export async function getUserFromAllUserStatusData(props, context) {
  const agencies = props?.agencies || context?.user?.agencies;

  const all = await Promise.allSettled(
    agencies.map(async ({ agencyId, userId, userIdType }) => ({
      agencyId,
      userIdType,
      ...(await context.datasources.getLoader("user").load({
        userId,
        agencyId,
      })),
    }))
  );

  const users = [...all.map((u) => u?.value)];

  // remove undefined fields
  users.forEach((obj) =>
    Object.keys(obj).forEach((key) => obj[key] === undefined && delete obj[key])
  );

  // Split data into CPR/LOCAL
  const cpr = users.filter((u) => u.userIdType === "CPR");
  const locals = users.filter((u) => u.userIdType === "LOCAL");

  // Prioritize CPR over LOCAL data
  return Object.assign({}, ...locals, ...cpr);
}

/**
 *
 * Function to find agencyId from branchId
 *
 * @param {string} branchId
 * @param {object} context
 * @returns {string} agencyId
 */

export async function getAgencyIdByBranchId(branchId, context) {
  const loader = context?.getLoader || context?.datasources?.getLoader;

  // get AgencyId from used branchId
  const result = (await loader("library").load({ branchId })).result?.[0];

  // return agencyId
  const agencyId = result?.agencyId;

  //  Return branchId instead of agencyId if branch act independently
  if (branchIsIndependent(branchId)) {
    return branchId;
  }

  return agencyId;
}
