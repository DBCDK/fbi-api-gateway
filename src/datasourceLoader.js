import DataLoader from "dataloader";

import { log } from "dbc-node-logger";
import { withRedis, clearRedis } from "./datasources/redis.datasource";
import { createFetchWithConcurrencyLimit } from "./utils/fetchWithLimit";
import { getFilesRecursive } from "./utils/utils";
import config from "./config";

/*
  This is how we time parallel requests to a single datasource.
  
  Request A
  ------------
          Request B
          ------------------
                                      Request C
                                      -----------
  Total AB
  --------------------------
  
  The totaltime for the requests should NOT be:
  Request A + Request B + Request C
  
  But instead:
  Total AB + Request C
*/

export const trackMe = () => {
  return {
    start: function (uuid) {
      this.uuid = uuid;
      return this;
    },
    createTracker: function (name) {
      // If this datasource has not been tracked yet, initialize its tracking object.
      if (!this.trackObject[name]) {
        this.trackObject[name] = {
          overlappingRequestCount: 0, // Number of ongoing requests for this datasource.
          count: 0, // Total number of elements retrieved from datasource
          started: 0, // Timestamp when the first concurrent request starts.
          time: 0, // Total time taken for all requests so far.
          sequentialTime: 0, // If requests were to be run sequentially, this is the total time
          batches: 0, // Dataloader batches multiple requests, into a single request when possible
        };
      }
    },

    // Function to indicate the beginning of a datasource call.
    begin: function (name, count) {
      this.createTracker(name);

      const beginTime = performance.now();

      const trackObject = this.trackObject[name];

      // Increment the count of ongoing requests.
      trackObject.overlappingRequestCount++;

      // Increment total number of elements retrieved from datasource
      trackObject.count += count;
      trackObject.batches++;

      // Function to indicate the end of an datasource call.
      return function end() {
        const endTime = performance.now();
        const duration = Math.round(endTime - beginTime);
        // Decrement the count of ongoing requests.
        trackObject.overlappingRequestCount--;

        // If all ongoing requests are complete, calculate the time taken.
        if (trackObject.overlappingRequestCount === 0) {
          trackObject.time += duration;
        }

        trackObject.sequentialTime += duration;
      };
    },

    cacheMiss: function (name, count) {
      this.createTracker(name);

      this.trackObject[name].cacheMiss =
        (this.trackObject[name].cacheMiss || 0) + count;
    },
    uuid: null,
    trackObject: {},
    getMetrics: function () {
      const res = {};

      Object.entries(this.trackObject).forEach(([key, val]) => {
        res[key] = {
          count: val.count,
          time: val.time,
          cacheMiss: val.cacheMiss,
          avg_ms: Math.round(val.sequentialTime / val.count),
        };
      });

      return res;
    },
  };
};

// Find all datasources in src/datasources
export const datasources = getFilesRecursive(`${__dirname}/datasources`)
  .map((file) => {
    if (!file.path.endsWith(".datasource.js")) {
      return;
    }
    const { load, options, batchLoader, status, testLoad } = require(file.path);
    if (!load && !batchLoader) {
      return;
    }

    // Extract datasource name from filename
    const name = file.file.replace(".datasource.js", "");

    return {
      testLoad,
      load,
      batchLoader,
      name,
      options,
      statusChecker: status ? () => status(load) : null,
    };
  })
  .filter((func) => !!func);

const nameToDatasource = {};
datasources.forEach((ds) => (nameToDatasource[ds.name] = ds));

log.debug(
  `found ${datasources.length} datasources, ${datasources
    .map((datasource) => datasource.name)
    .join(", ")}`
);

/**
 * This will initialize a single dataloader.
 * It makes sure a context (with a concurrency limited fetcher), is injected
 * into the batchloader function.
 *
 * It sets up Redis for the dataloader
 *
 * It collects performance metrics for the datasource.
 */
function setupDataloader(
  { name, load, options, batchLoader, testLoad },
  context
) {
  const testModeActivated = context?.testUser && testLoad;
  const loadImpl = testModeActivated ? testLoad : load;

  let batchLoaderWithContext = batchLoader
    ? (keys) => batchLoader(keys, context)
    : (keys) => Promise.all(keys.map((key) => loadImpl(key, context)));

  if (!testModeActivated && options?.redis?.prefix && options?.redis?.ttl) {
    batchLoaderWithContext = withRedis(batchLoaderWithContext, {
      ...options.redis,
      ...context,
      datasourceName: name,
    });
  }

  async function batchLoaderWithTiming(keys) {
    const end = context?.track?.begin?.(name, keys.length);

    try {
      return await batchLoaderWithContext(keys);
    } catch (err) {
      // Wraps the load call in a try catch, in order to capture the
      // stack trace before dataloader is called. Otherwise we would
      // be limited to see the stack trace from the data sources, but
      // not the GraphQL resolvers.
      // https://brunoscheufler.com/blog/2021-07-25-improving-dataloader-stack-traces
      if (err instanceof Error && err.stack) {
        const currentStack = { stack: "" };
        Error.captureStackTrace(currentStack, batchLoaderWithTiming);
        err.stack += currentStack.stack.replace("Error", "\nBefore DataLoader");
      }
      throw err;
    } finally {
      end?.();
    }
  }

  const loader = new DataLoader(batchLoaderWithTiming, {
    // If key is an object, we stringify
    // to make it useful as a cache key
    cacheKeyFn: (key) => (typeof key === "object" ? JSON.stringify(key) : key),
    maxBatchSize: 100,
  });

  if (options?.redis?.prefix) {
    loader.clearRedis = (key) => clearRedis(options?.redis?.prefix, key);
  }

  return {
    loader,
  };
}

/**
 * Will instantiate dataloaders from datasources.
 * This should be done for every incoming GraphQL request
 */
export default function createDataLoaders(uuid, testUser, accessToken) {
  const result = {};
  const track = trackMe().start(uuid);

  result.trackingObject = track;
  const fetchWithConcurrencyLimit = createFetchWithConcurrencyLimit(
    config.fetchConcurrencyLimit
  );
  // Gets a loader by name.
  // A loader will be initialized first time it is called
  function getLoader(name) {
    if (!result[name]) {
      result[name] = setupDataloader(nameToDatasource[name], {
        getLoader,
        track,
        fetch: (url, options) => fetchWithConcurrencyLimit(url, options, name),
        trackingId: uuid,
        testUser,
        accessToken,
      })?.loader;
    }
    return result[name];
  }
  result.getLoader = getLoader;
  return result;
}

/**
 * For testing
 */
export function createMockedDataLoaders() {
  const mockedDatasources = {};
  getFilesRecursive("./src/datasources/mocked")
    .filter(
      (file) =>
        file.path.endsWith("datasource.mocked.js") && require(file.path).load
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.mocked.js", ""),
      load: require(file.path).load,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDatasources[loader.name] = loader;
    });

  mockedDatasources.getLoader = (name) => mockedDatasources[name];

  return mockedDatasources;
}

export function createTestUserDataLoaders() {
  const mockedDataLoaders = createMockedDataLoaders();
  getFilesRecursive("./src/datasources")
    .filter(
      (file) =>
        file.path.endsWith("datasource.js") && require(file.path).testLoad
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.js", ""),
      load: require(file.path).testLoad,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDataLoaders[loader.name] = loader;
    });

  return mockedDataLoaders;
}
