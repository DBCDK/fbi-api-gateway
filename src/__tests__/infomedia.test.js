import { createMockedDataLoaders } from "../datasourceLoader";
import { performTestQuery } from "../utils/utils";

test.skip("InfomediaContent, 2 articles", async () => {
  const result = await performTestQuery({
    query: `
        query ($pid: String!) {
          infomediaContent(pid:$pid){
            id
            html
          }
        }
        `,
    variables: { pid: "870971-avis:34591016" },
    context: { datasources: createMockedDataLoaders() },
  });

  expect(result).toMatchSnapshot();
});

test("InfomediaContent, single article", async () => {
  const result = await performTestQuery({
    query: `
        query ($id: String!) {
  infomedia(id: $id) {
    error
    article {
      id
      html
    }
  }
}
        `,
    variables: { id: "e2a3ae5c" },
    context: {
      smaug: {},
      user: { userId: "fisk" },
      accessToken: "qwerty",
      datasources: createMockedDataLoaders(),
    },
  });

  expect(result).toMatchSnapshot();
});

test.skip("InfomediaContent, multiple reviews", async () => {
  const result = await performTestQuery({
    query: `
        query ($pid: String!) {
          infomediaContent(pid:$pid){
            id
            html
          }
        }
        `,
    variables: { pid: "870971-anmeld:47413281" },
    context: { datasources: createMockedDataLoaders() },
  });

  expect(result).toMatchSnapshot();
});

test.skip("Manifestation -> onlineAccess returns InfomediaReference", async () => {
  const result = await performTestQuery({
    query: `
        query ($pid: String!) {
         manifestation(pid:$pid){
            pid
            onlineAccess {
                ... on InfomediaReference {
                  infomediaId
                  
                }
            }
          }
        }
        `,
    variables: { pid: "870971-anmeld:47413281" },
    context: {
      accessToken: "qwerty",
      datasources: createMockedDataLoaders(),
    },
  });

  expect(result).toMatchSnapshot();
});
