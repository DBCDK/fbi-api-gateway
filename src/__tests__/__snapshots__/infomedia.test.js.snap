// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfomediaContent, 2 articles 1`] = `
Object {
  "data": Object {
    "infomediaContent": Array [
      Object {
        "html": "<div class=\\"infomedia_HeadLine\\">Afghanistan: NATO bruger hvid fosfor i stort omfang mod Taleban</div>
<div class=\\"infomedia_SubHeadLine\\">Afghanistan: NATO bruger hvid fosfor i stort omfang mod Taleban</div>
<div class=\\"infomedia_ByLine\\">Charlotte Aagaard, Anton Geist, Sebastian Gjerding, Emil Rottbøll, ,, ,  Og</div>
<div class=\\"infomedia_DateLine\\">19. April 2011</div>
<div class=\\"infomedia_paper\\">Information</div>
<div class=\\"infomedia_hedline\\">NATO-koalitionen skyder med hvid fosfor i Afghanistan i langt større omfang, end det hidtil har været kendt. Flere efterretningsrapporter tyder på, at det napalmlignende stof bruges ulovligt som angrebsvåben. Også danske styrker har brugt hvid fosfor i beboede områder</div>
<div class=\\"infomedia_text\\"><p id=\\"p1\\">Nye lækkede efterretningsrapporter viser et overraskende stort forbrug af hvid fosfor i krigen mod Taleban.</p><p id=\\"p2\\">De knap 15.000 efterretningsrapporter fra koalitionen i Afghanistan, som Information er i besiddelse af, viser eksempler på, at NATO-koalitionens styrker ( ISAF) har skudt 20, 30 og 50 runder af det napalmlignende stof mod ét enkelt oprørsmål.</p><p id=\\"p3\\">Normalt bruges hvid fosfor til at oplyse et krigsområde om natten, udlægge røgslør eller markere et bombemål.</p><p id=\\"p4\\">Men der er adskillige efterretningsrapporter, som tyder på, at hvid fosfor anvendes som angrebsvåben mod Taleban og deres allierede.</p><p id=\\"p5\\">Ifølge FN's Våbenkonvention er bomber med hvid fosfor ulovlige at bruge i krig, når det sker i »tætbebyggede områder«, eller hvis der er risiko for at ramme civile.</p><p id=\\"p6\\">Hvid fosfor som angreb I en af de lækkede rapporter fremgår det, at amerikanske styrker 19. juni 2008 i det østlige Afghanistan har fået øje på en fjendtlig oprørsstilling mellem kl. 9.59 og 12.22 afghansk tid.</p><p id=\\"p7\\">»Som svar affyrede Task Force Eagle 30 gange 120 mm hvid fosfor på denne koordinat og smed to GBU 31 ( bomber) i nærheden af samme koordinat,« hedder det i ISAF-rapport # 06-862. ISAF bekræfter brugen af fosfor som angrebsvåben, men afviser, at det bruges ulovligt.</p><p id=\\"p8\\">Den omfattende brug af hvid fosfor bekymrer menneskerettighedsorganisationer, der mener, at det er problematisk i en krig, der ofte foregår i beboede områder som den Grønne Zone, hvor de danske soldater er udstationeret.</p><p id=\\"p9\\">Peter Vedel Kessing, seniorforsker og folkeretsekspert ved Institut for Menneskerettigheder, mener, at det bør undgås: »I det omfang, hvid fosfor anvendes direkte mod kombattanter i Afghanistan, kan der efter min vurdering være tale om unødvendig og overflødig skadevirkning i forhold til de opnåede militære mål.« De nye dokumenter omfatter også eksempler på, at danske styrker har anvendt hvid fosfor i beboede områder.</p><p id=\\"p10\\">Militærjurist Rolf Verge fra Hærens Operative Kommando bekræfter, at de danske styrker anvender hvid fosfor, men han forsikrer, at der udelukkende er tale om lovlige formål.</p><p id=\\"p11\\">CAA@INFORMATION.DK</p></div>
<div class=\\"infomedia_logo\\">
<img src=\\"infomedia_logo.gif\\" alt=\\"logo\\"/>
<p>Alt materiale i Infomedia er omfattet af lov om ophavsret og må ikke kopieres uden særlig tilladelse.</p>
</div>",
        "id": "e2a3ae5c",
      },
      Object {
        "html": "<div class=\\"infomedia_logo\\">
<img src=\\"infomedia_logo.gif\\" alt=\\"logo\\"/>
<p>Alt materiale i Infomedia er omfattet af lov om ophavsret og må ikke kopieres uden særlig tilladelse.</p>
</div>",
        "id": "e2a3ae63",
      },
    ],
  },
}
`;

exports[`InfomediaContent, multiple reviews 1`] = `
Object {
  "data": Object {
    "infomediaContent": Array [
      Object {
        "html": "<div class=\\"infomedia_HeadLine\\">Slap krimi til at lægge under træet</div>
<div class=\\"infomedia_SubHeadLine\\">Slap krimi til at lægge under træet</div>
<div class=\\"infomedia_ByLine\\">,Nanna Frank Rasmussen</div>
<div class=\\"infomedia_DateLine\\">16. November 2019</div>
<div class=\\"infomedia_paper\\">Politiken</div>
<div class=\\"infomedia_hedline\\">Efter en afstikker til USA med den spændstige bedemandstrilogi vender Sara Blædel hjem til sine kendte karakterer Louise Rick og Camilla Lind.</div>
<div class=\\"infomedia_text\\"><p id=\\"p1\\">Sara Blædel vækker for tiende gang sine kendte og elskede figurer politiassistenten Louise Rick og journalisten Camilla Lind til live. De mange trofaste læsere af Blædels krimier skal nok blive begejstrede bare ved tanken om, at der nu er nyt at høre om Rick og Linds tumultariske kærlighedsliv og forældreskabsudfordringer.</p><p id=\\"p2\\">Det er trods alt de to karakterers oplevelser, der har givet liv til Sara Blædels populære hverdagskrimier, der har haft en stor fanskare siden ' Grønt støv' og ' Kald mig prinsesse' udkom i hhv. 2004 og 2005.</p><p id=\\"p3\\">Efter ' Pigen under træet' må man dog spørge, om der snart kan vrides mere ud af den serie. Skal bogmarkedet og fans tvangsfodres med mere uinspireret venstrehåndskrimi, når nu Blædel har vist, at hun kan nå et helt andet og højere niveau? Hendes sidste tre bøger spruttede nemlig af en anderledes fortælleglæde end den, man nu må lade sig nøjes med.</p><p id=\\"p4\\">Den spændstigt spændende udviklingsog roadromantrilogi om Ilka og hendes søstre, nu kendt som bedemandstrilogien, foregik som bekendt i USA, og måske var det så enkelt, at det fysiske flyt gav luft både under fantasiens og sprogets vinger. I så fald ville man ønske, at forfatteren var forblevet udlandsdansker.</p><p id=\\"p5\\">For selv hvis Sara Blædel vendte hjem til sine gamle papirlegekammerater med begejstring, smitter det mildest talt hverken af på sprog eller plot, der er både slapt og søvndyssende.</p><p id=\\"p6\\"><br /><em style=\\"bold\\">Med på lejrskole</em></p><p id=\\"p7\\">Den nyeste bog i serien om Rick og Lind drypper af privat hjemstavnsnostalgi, for med ' Pigen under træet' er vi endnu en gang tilbage i Roskilde, Osted og omegn, hvor forfatteren selv har boet. Desuden finder den ene af bogens forbrydelser sted på en lejrskole på Bornholm i 1995. Også den skoletur er kalkeret over forfatterens (og hendes læseres) oplevelser.</p><p id=\\"p8\\">I 1995 forvilder en ung pige sig ind i en klippegrotte i Ekkodalen, bange og forfulgt.</p><p id=\\"p9\\">Knap 25 år senere finder man hendes lig.</p><p id=\\"p10\\">Samtidig forsvinder Louise Ricks svigerinde, og det får hendes mand, Ricks bror, Mikkel, til at forsøge at begå selvmord.</p><p id=\\"p11\\">Det er ikke første gang, at svigerinden har forladt familien, men det virker alligevel mærkværdigt, at hun ikke har taget nogen stor bagage med sig.</p><p id=\\"p12\\">Snart anklages Mikkel for selv at stå bag sin kones forsvinden. Ligger der mon en forbrydelse bag? Rick kaldes hjem fra sin jordomrejse med kæresten, Eik, og plejesønnen, Jonas, og snart er hun og veninden Camilla involveret i opklaringen af både den nye og den gamle sag.</p><p id=\\"p13\\">Handler det om en ' familietragedie', som medierne ynder at kalde mord og koldblodige likvideringer inden for familiens rammer, eller er der noget helt andet indviklet på spil? Der er så mange spørgsmål, der hober sig op i ' Pigen under træet', men desværre ikke på den gode genredefinerede måde. For bikaraktererne lider under en eklatant mangel på psykologisk dybde og troværdighed.</p><p id=\\"p14\\">Mikkels svigermor vender for eksempel på en tallerken efter en let pædagogisk samtale med Louise Rick. Svigermoren tror nemlig, at det er Mikkel, der står bag sin kones, hendes datters, forsvinden, for hun har hørt (!), at det oftest er nærtstående, der er de skyldige i den slags ' familietragedier' (vi ved først meget sent, hvad der egentlig er sket). Men Louise får hende altså omvendt nærmest ved at knipse med fingrene.</p><p id=\\"p15\\">Svigermor mener desuden, at selvmordsforsøget er afledt af skyldfølelse, men Louise forklarer, at det skyldes, at han er blevet forladt (det er nemlig ikke første gang): »Hans hjerte er knust. Han føler hjertesorg«, som hendes poetiske forklaring lyder.</p><p id=\\"p16\\">Da svigermoren bliver forsikret om, at Rick og Lind nok skal finde ud af, hvad der er sket, sadler hun straks om.</p><p id=\\"p17\\">»Det var måske også lidt overilet at tro, at Mikkel havde gjort noget slemt, han er jo en sød dreng. Men jeg ved, at der er sket min datter noget«, lyder replikken, der med et får 150 siders forudgående mistro og alvorlige anklager til at forsvinde som dug fra solen.</p><p id=\\"p18\\"><br /><em style=\\"bold\\">Julegavebestseller</em></p><p id=\\"p19\\">Lige så nemt kommer Mikkel ud af en kulsort depression, der har drænet ham for enhver livslyst. Og da hele plottets baggrund skal vikles op og ud, sker det i et hæsblæsende nonchalant tempo. Sådan var det bare. Fordi. Punktum. Det er enten dovent eller direkte provokerende nedladende over for det læsende publikum, at Blædel ikke er mere interesseret i at udfolde ondskabens kerne og bevæggrunde.</p><p id=\\"p20\\">Der er tilløb til, at læseren kan lægge bogen fra sig med overordnet tematiske betragtninger i hukommelsesbanken, men det er kun, hvis man stædigt insisterer på at komme ud af refleksionens starthuller. Man kan med lidt velvillighed sige, at ' Pigen under træet' kredser om det at blive forladt, udstødt, at føle sig uelsket og fravalgt. Noget, som denne bog, udgivelsestidspunktet og fanbasens størrelse taget i betragtning ikke kommer til at opleve.</p><p id=\\"p21\\">Først og fremmest føles det nemlig, som om bogen er blevet til, fordi forfatter og forlag gerne ville præsentere en oplagt julegavebestseller. Man kunne derfor sagtens have kaldt den ' Bogen under juletræet' i stedet.</p><p id=\\"p22\\">nanna.rasmussen@pol.dk </p><p id=\\"p23\\"><br /><em style=\\"bold\\">Fakta: ANMELDERPLUK Det skrev Politiken om de 3 seneste i Louise Rick-serien</em></p><p id=\\"p24\\">' Kvinden de meldte savnet' (2014) »Nej, der er ikke meget rå vold og vildskab over Blædel, men hun forstår at fortælle en rolig og paradoksalt nok tempofyldt historie om død og kærlighed«. Mette Højbjerg ' Dødesporet' (2013) »Det er befriende at læse om religiøs ekstremisme og vold, uden at det handler om islamister og Mellemøsten.</p><p id=\\"p25\\">Også derfor kører det godt for Sara Blædel i den ottende bog om opdageren Louise Rick«. Mette Højbjerg ' De glemte piger' (2011) »Sara Blædel har med ' De glemte piger' ikke bare fundet på en rigtig god historie, hun har også skruet plottet suverænt sammen, så overraskelserne kommer på de helt rigtige tidspunkter, og uhyggen indfinder sig i stadig stigende grad«. Mette Højbjerg.</p></div>
<div class=\\"infomedia_logo\\">
<img src=\\"infomedia_logo.gif\\" alt=\\"logo\\"/>
<p>Alt materiale i Infomedia er omfattet af lov om ophavsret og må ikke kopieres uden særlig tilladelse.</p>
</div>",
        "id": "e770b915",
      },
    ],
  },
}
`;

exports[`InfomediaContent, single article 1`] = `
Object {
  "data": Object {
    "infomedia": Object {
      "article": Object {
        "html": "<div class=\\"infomedia_HeadLine\\">Afghanistan: NATO bruger hvid fosfor i stort omfang mod Taleban</div>
<div class=\\"infomedia_SubHeadLine\\">Afghanistan: NATO bruger hvid fosfor i stort omfang mod Taleban</div>
<div class=\\"infomedia_ByLine\\">Charlotte Aagaard, Anton Geist, Sebastian Gjerding, Emil Rottbøll, ,, ,  Og</div>
<div class=\\"infomedia_DateLine\\">19. April 2011</div>
<div class=\\"infomedia_paper\\">Information</div>
<div class=\\"infomedia_hedline\\">NATO-koalitionen skyder med hvid fosfor i Afghanistan i langt større omfang, end det hidtil har været kendt. Flere efterretningsrapporter tyder på, at det napalmlignende stof bruges ulovligt som angrebsvåben. Også danske styrker har brugt hvid fosfor i beboede områder</div>
<div class=\\"infomedia_text\\"><p id=\\"p1\\">Nye lækkede efterretningsrapporter viser et overraskende stort forbrug af hvid fosfor i krigen mod Taleban.</p><p id=\\"p2\\">De knap 15.000 efterretningsrapporter fra koalitionen i Afghanistan, som Information er i besiddelse af, viser eksempler på, at NATO-koalitionens styrker ( ISAF) har skudt 20, 30 og 50 runder af det napalmlignende stof mod ét enkelt oprørsmål.</p><p id=\\"p3\\">Normalt bruges hvid fosfor til at oplyse et krigsområde om natten, udlægge røgslør eller markere et bombemål.</p><p id=\\"p4\\">Men der er adskillige efterretningsrapporter, som tyder på, at hvid fosfor anvendes som angrebsvåben mod Taleban og deres allierede.</p><p id=\\"p5\\">Ifølge FN's Våbenkonvention er bomber med hvid fosfor ulovlige at bruge i krig, når det sker i »tætbebyggede områder«, eller hvis der er risiko for at ramme civile.</p><p id=\\"p6\\">Hvid fosfor som angreb I en af de lækkede rapporter fremgår det, at amerikanske styrker 19. juni 2008 i det østlige Afghanistan har fået øje på en fjendtlig oprørsstilling mellem kl. 9.59 og 12.22 afghansk tid.</p><p id=\\"p7\\">»Som svar affyrede Task Force Eagle 30 gange 120 mm hvid fosfor på denne koordinat og smed to GBU 31 ( bomber) i nærheden af samme koordinat,« hedder det i ISAF-rapport # 06-862. ISAF bekræfter brugen af fosfor som angrebsvåben, men afviser, at det bruges ulovligt.</p><p id=\\"p8\\">Den omfattende brug af hvid fosfor bekymrer menneskerettighedsorganisationer, der mener, at det er problematisk i en krig, der ofte foregår i beboede områder som den Grønne Zone, hvor de danske soldater er udstationeret.</p><p id=\\"p9\\">Peter Vedel Kessing, seniorforsker og folkeretsekspert ved Institut for Menneskerettigheder, mener, at det bør undgås: »I det omfang, hvid fosfor anvendes direkte mod kombattanter i Afghanistan, kan der efter min vurdering være tale om unødvendig og overflødig skadevirkning i forhold til de opnåede militære mål.« De nye dokumenter omfatter også eksempler på, at danske styrker har anvendt hvid fosfor i beboede områder.</p><p id=\\"p10\\">Militærjurist Rolf Verge fra Hærens Operative Kommando bekræfter, at de danske styrker anvender hvid fosfor, men han forsikrer, at der udelukkende er tale om lovlige formål.</p><p id=\\"p11\\">CAA@INFORMATION.DK</p></div>
<div class=\\"infomedia_logo\\">
<img src=\\"infomedia_logo.gif\\" alt=\\"logo\\"/>
<p>Alt materiale i Infomedia er omfattet af lov om ophavsret og må ikke kopieres uden særlig tilladelse.</p>
</div>",
        "id": "e2a3ae5c",
      },
      "error": null,
    },
  },
}
`;

exports[`Manifestation -> onlineAccess returns InfomediaReference 1`] = `
Object {
  "data": Object {
    "manifestation": Object {
      "onlineAccess": Array [
        Object {
          "infomediaId": "e2a3ae5c",
        },
        Object {
          "infomediaId": "e2a3ae63",
        },
      ],
      "pid": "870971-anmeld:47413281",
    },
  },
}
`;
